services/payments/service.go

```diff
@@ -0,0 +1,180 @@
+	if !ok {
+		return errors.New("template secret id for enclave decrypt key not found on context")
+	}
+
+	smClient := secretsmanager.NewFromConfig(cfg)
+	o, err := smClient.GetSecretValue(ctx, &secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(templateSecretID),
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to get key policy template from secrets manager: %w", err)
+	}
+
+	if o.SecretString == nil {
+		return errors.New("secret is not defined in secrets manager")
+	}
+
+	keyPolicy := *o.SecretString
+	keyPolicy = strings.ReplaceAll(keyPolicy, "<IMAGE_SHA384>", imageSha384)
+	keyPolicy = strings.ReplaceAll(keyPolicy, "<PCR0>", pcr0)
+	keyPolicy = strings.ReplaceAll(keyPolicy, "<PCR1>", pcr1)
+	keyPolicy = strings.ReplaceAll(keyPolicy, "<PCR2>", pcr2)
+
+	client := kms.NewFromConfig(cfg)
+
+	input := &kms.CreateKeyInput{
+		Policy: aws.String(keyPolicy),
+	}
+
+	result, err := awsutils.MakeKey(ctx, client, input)
+	if err != nil {
+		return fmt.Errorf("failed to make key: %w", err)
+	}
+
+	s.kmsDecryptKeyArn = *result.KeyMetadata.KeyId
+	return nil
+}
+
+// NewService creates a service using the passed datastore and clients configured from the environment
+func NewService(ctx context.Context) (context.Context, *Service, error) {
+	var logger = logging.Logger(ctx, "payments.NewService")
+
+	service := &Service{
+		baseCtx:   ctx,
+		secretMgr: &awsClient{},
+	}
+
+	if err := service.configureKMSKey(ctx); err != nil {
+		logger.Fatal().Msg("could not create kms secret decryption key")
+	}
+
+	if err := service.configureDatastore(ctx); err != nil {
+		logger.Fatal().Msg("could not configure datastore")
+	}
+
+	// setup our custodian integrations
+	upholdCustodian, err := provider.New(ctx, provider.Config{Provider: provider.Uphold})
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to create uphold custodian")
+		return ctx, nil, fmt.Errorf("failed to create uphold custodian: %w", err)
+	}
+	geminiCustodian, err := provider.New(ctx, provider.Config{Provider: provider.Gemini})
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to create gemini custodian")
+		return ctx, nil, fmt.Errorf("failed to create gemini custodian: %w", err)
+	}
+	bitflyerCustodian, err := provider.New(ctx, provider.Config{Provider: provider.Bitflyer})
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to create bitflyer custodian")
+		return ctx, nil, fmt.Errorf("failed to create bitflyer custodian: %w", err)
+	}
+
+	service.custodians = map[string]provider.Custodian{
+		provider.Uphold:   upholdCustodian,
+		provider.Gemini:   geminiCustodian,
+		provider.Bitflyer: bitflyerCustodian,
+	}
+
+	return ctx, service, nil
+}
+
+// DecryptBootstrap - use service keyShares to reconstruct the decryption key
+func (s *Service) DecryptBootstrap(ctx context.Context, ciphertext []byte) (map[appctx.CTXKey]interface{}, error) {
+	// combine the service configured key shares
+	key, err := shamir.Combine(s.keyShares)
```

Comment:
Where are these keyShares getting decrypted and instantiated at within the service? I'm not seeing this happen in the function above so a bit confused how we're pulling them out of the S3 bucket and decrypting them here.